File: typescript-workshop/part-4-behavior/src/service/GatewayShippingService.ts
import { Order } from '../domain/Order';
import { ShippingConfirmation } from '../domain/ShippingConfirmation';
import { ShippingItem } from '../domain/ShippingItem';
import { ShippingRequest } from '../domain/ShippingRequest';
import { ShippingGateway } from '../gateway/ShippingGateway';
import { ShippingService } from './ShippingService';

export class GatewayShippingService implements ShippingService {
  constructor(private readonly shippingGateway: ShippingGateway) {}

  createShipment(order: Order): ShippingConfirmation {
    const items = order.items.map((item) => new ShippingItem(item.productId, item.quantity));
    const request = new ShippingRequest(order.id, order.customerId, items);
    return this.shippingGateway.createShipment(request);
  }

  cancelShipment(trackingNumber: string): boolean {
    return this.shippingGateway.cancelShipment(trackingNumber);
  }
}

File: csharp-workshop/Part2.LightMocks/Service/GatewayShippingService.cs
using Part2.LightMocks.Domain;
using Part2.LightMocks.Gateway;

namespace Part2.LightMocks.Service;

public class GatewayShippingService : IShippingService
{
    private readonly IShippingGateway _shippingGateway;

    public GatewayShippingService(IShippingGateway shippingGateway)
    {
        _shippingGateway = shippingGateway;
    }

    public ShippingConfirmation CreateShipment(Order order)
    {
        var request = new ShippingRequest(
            order.Id,
            order.CustomerId,
            order.Items.Select(item => new ShippingItem(item.ProductId, item.Quantity)).ToList()
        );
        return _shippingGateway.CreateShipment(request);
    }

    public bool CancelShipment(string trackingNumber)
    {
        return _shippingGateway.CancelShipment(trackingNumber);
    }
}
